Workarounds for Linux build errors

elsiehupp opened the pull request:
Added workarounds for the build issues I experienced several days back:
- `node-pre-gyp` sometimes calls an invalid URL and needs Python to build `node-sqlite3` from source
- `npm install` can be extremely fussy. `npm cache clean --force` doesn't help, so just nuking the directory and rebuilding it can make `npm` behave. (`npm rebuild` may be another alternative.)

<!--

Please prefix the title with the platform you are targetting:

Here are some examples of good titles:

- Desktop: Resolves #123: Added new setting to change font
- Mobile, Desktop: Fixes #456: Fixed config screen error
- All: Resolves #777: Made synchronisation faster

And here's an explanation of the title format:

- "Desktop" for the Windows/macOS/Linux app (Electron app)
- "Mobile" for the mobile app (or "Android" / "iOS" if the pull request only applies to one of the mobile platforms)
- "CLI" for the CLI app

If it's two platforms, separate them with commas - "Desktop, Mobile" or if it's for all platforms, prefix with "All".

If it's not related to any platform (such as a translation, change to the documentation, etc.), simply don't add a platform.

Then please append the issue that you've addressed or fixed. Use "Resolves #123" for new features or improvements and "Fixes #123" for bug fixes.

AND PLEASE READ THE GUIDE: https://github.com/laurent22/joplin/blob/dev/CONTRIBUTING.md

-->

laurent22 commented:
Thanks for the doc update, but I'm not sure the first advise would make a big difference in most cases. There's a command `npm run clean` that might be more useful.

As for the second issue `node-pre-gyp WARN Tried to download(403)`, it's not an error, it just means that SQLite is going to be built from source. Maybe it's part of a larger issue, that we should check if Python and other tools are present before the build starts, but there's no easy way to do this as far as I know.

elsiehupp commented:
I had a much more elaborate draft put together for explaining the Python error, but I scrapped it in favor of brevity. Basically, under certain circumstances, as you know, `node-sqlite3` needs to be compiled from source, and when that happens, Python is a dependency. If you don't have `python-is-python2` or `python-is-python3` installed, then this is a fatal error that prevents `npm i` from fulling completing. You could probably figure out a way of handling this more elegantly, but if it isn't immediately obvious to someone encountering this error, then, yes, `sudo apt install python` is the way to resolve it. You could list Python elsewhere as a build requirement, but it's only a sometimes-maybe build requirement.

elsiehupp commented:
I went ahead and added Python (i.e. `python-is-python2` or `python-is-python3`) to the dependency `apt` command. I also added a reference to not using Node 15, because I had varied and inscrutable problems on my Mac when I had Node 15 installed that went away when I downgraded to Node 14 LTS.

laurent22 commented on BUILD.md:
```diff
@@ -20,11 +20,11 @@ There are also a few forks of existing packages under the "fork-*" name.
 ## Required dependencies
 
-- Install node 10+ - https://nodejs.org/en/
+- Install node 10+ (but not node 15) - https://nodejs.org/en/
```
If the app cannot be built under Node 15, it's more a bug that should be fixed (and that I'll fix when I upgrade) rather than something to add to the doc, so please could you remove this part?

elsiehupp replied:
Oh, okay. I mean, it's still perhaps worth mentioning in some way or another if Joplin hasn't been tested yet under Node 15. I specifically ran into the issues on Mac, for whatever that's worth.

laurent22 commented:
Looks good now, thanks for the update! For Node 15, I will probably give it a try in the coming days so hopefully I can fix whatever issue there is.
